core/libs/commonwealth/commonwealth/utils/settings.py

```diff
@@ -0,0 +1,197 @@
+from typing import Any, Dict, Optional
+from loguru import logger
+import appdirs
+import os
+import pathlib
+import json
+import importlib.util
+import time
+
+class NoSettingsFile(ValueError):
+    """Settings file does not exist."""
+
+class BadSettingsFile(ValueError):
+    """Settings file is not valid."""
+
+class SettingsFromTheFuture(ValueError):
+    """Settings file version is from a newer version of the service."""
+
+class OldSettings(ValueError):
+    """Settings file version is old."""
+
+class MigrationFail(RuntimeError):
+    """Could not apply migration."""
+
+class BadAttributes(BadSettingsFile):
+    """Attributes on settings file are not valid for current version."""
+
+
+class Settings:
```

Comment:
That goes closer to what we have discussed, but there are some things that are a bit dangerous. The access of settings variable via a string argument makes it unsafe to access data and more error prone, that's also a problem related to the entire dict logic.
With a serialization and deserialization library this is fixed, we also ensure that the settings variables are known with static analysis tools.
I wrote a simple example that demonstrates such idea here:
https://github.com/patrickelectric/python_settings_magic/blob/master/settings.py

Reply:
Yup, that access directly by dicts is less than ideal. Glad you took a look at that.

In your example we are assuming that variables are just added, not removed, from version to version, or they can also be removed by just not migrating them? In this case, static analysis tools are still going to think they are there, since they are inherited from the parent class, right?

Reply:
If so, a solution would be to simple re-declare an "attributes" variable in each version, which would store the entire attributes structure.

Reply:
You can just create another class and replace the migrate without super:
```python
class SettingsV3(SettingsV1):
    VERSION = 3
    new_variable_tresh = pykson.IntegerField(default_value=42)

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.VERSION = SettingsV3.VERSION

    def migrate(self, data: dict):
        if data["VERSION"] < SettingsV3.VERSION:
            SettingsV2().migrate(data)

        data["VERSION"] = SettingsV3.VERSION
        data.pop("new_variable_dosh")
```

Reply:
Didn't get where it solves the problem...

Reply:
Check the repository for a working example.